### PR TITLE
Fix character set mapping for `utf8` and use it in schema-tracking tables

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR builds on the changes made in https://github.com/vitessio/vitess/pull/14904.

In that PR, we used `utf8mb3` as the character set and `utf8mb3_bin` for the collation of schema tracking tables since we wanted them to be case-sensitive. However, we realized that `utf8mb3` is deprecated in MySQL 5.7.

In this PR, we change the schema definitions to use `utf8` instead of `utf8mb3` and also make changes to Vitess to replicate the MySQL behavior where `utf8` maps to `utf8mb3` in MySQL 5.7 and `utf8mb4` in MySQL 8.0.

Tests have been added to verify that the schema-diff creates the correct tables in MySQL 5.7 and MySQL 8.0 and the upgrades and downgrades generate the correct alter statements too.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/collations/mysqlversion.go [hunk 1]
```diff
@@ -145,7 +152,7 @@ var globalVersionInfo = map[ID]struct {
 	80:   {alias: []collalias{{0b01111111, "cp850_bin", "cp850"}}, isdefault: 0b00000000},
 	81:   {alias: []collalias{{0b01111111, "cp852_bin", "cp852"}}, isdefault: 0b00000000},
 	82:   {alias: []collalias{{0b01111111, "swe7_bin", "swe7"}}, isdefault: 0b00000000},
-	83:   {alias: []collalias{{0b01111111, "utf8_bin", "utf8"}, {0b01111111, "utf8mb3_bin", "utf8mb3"}}, isdefault: 0b00000000},
+	83:   {alias: []collalias{{0b01111111, "utf8_bin", "utf8"}, {0b01111111, "utf8mb3_bin", "utf8mb3"}, {0b01111111, "utf8mb4_bin", "utf8mb4"}}, isdefault: 0b00000000},
```
Comment (GuptaManan100) [inline go/mysql/collations/mysqlversion.go:155]: I am not sure about this change.

**Review by GuptaManan100 — COMMENTED** [review]
I have made changes based on what I think is the correct way of doing this, but I could be wrong. I would love to have a review on this PR from @vmg or @dbussink to ensure that this is indeed the correct way.

File: go/mysql/collations/mysqlversion.go (hunk 1 above)
Comment (dbussink) [reply go/mysql/collations/mysqlversion.go:155]: This is a generated file based on the actual MySQL collations. So this change is wrong.

**Review by dbussink — CHANGES_REQUESTED** [review]
> In that PR, we used `utf8mb3` as the character set and `utf8mb3_bin` for the collation of schema tracking tables since we wanted them to be case-sensitive. However, we realized that `utf8mb3` is deprecated in MySQL 5.7.

I don't understand this. We didn't then afterwards realize. It was already known that it's deprecated, but since MySQL itself uses internally still `utf8mb3` for all table information (see `information_schema` usage and in Vitess all usage of https://github.com/vitessio/vitess/blob/e9c513e504928607b39a742ae393e4527cf94109/go/mysql/collations/env.go#L200-L208 for example). 

So there's no need for any follow up that I know of. 

> In this PR, we change the schema definitions to use `utf8` instead of `utf8mb3` and also make changes to Vitess to replicate the MySQL behavior where `utf8` maps to `utf8mb3` in MySQL 5.7 and `utf8mb4` in MySQL 8.0.

This is not true. `utf8` is still an alias for `utf8mb3` in MySQL 8.0. And given MySQL's new versioning scheme, I don't think it will ever change for 8.0 and any changes will happen in a newer version.

**Comment by vmg** [comment]
Yeah I also don't think it's correct to change the encoding for these schema tables. The `utf8mb3` is indeed deprecated but it's still being used in MySQL for these tables, so this change just breaks compatibility. Obviously it'd be neat to support emoji and extended codepoints in table names (would it tho?) but until MySQL also supports it, what's the point?

**Comment by GuptaManan100** [comment]
Thank-you both for looking at the PR! I have closed it in light of it not being a change that makes sense.

**Pull request closed without merging**